src/dns.c

```diff
@@ -2098,87 +2120,100 @@ hsk_dns_name_parse(
   uint8_t **data_,
   size_t *data_len_,
   const hsk_dns_dmp_t *dmp,
-  char *name
+  uint8_t *name
 ) {
+  // Local copies of message read pointers
   uint8_t *data = *data_;
   size_t data_len = *data_len_;
+  // Offset of `data` array being parsed
   int off = 0;
+  // Offset of `name` output array
   int noff = 0;
+  // Message read resume location after name read
   int res = 0;
-  int max = HSK_DNS_MAX_NAME;
+  // Pointer count
   int ptr = 0;
 
   for (;;) {
+    // Can not read past end of input array
     if (off >= data_len)
       return -1;
 
+    // Leading byte (size of label or compression pointer)
     uint8_t c = data[off];
-    off += 1;
 
-    if (c == 0x00)
+    // Only occurs at end of name (zero-length label aka ".")
+    if (c == 0x00) {
+      // Almost made it! Last byte exceeds max.
+      if (noff + 1 > HSK_DNS_MAX_NAME)
+        return -1;
+
+      if (name)
+        name[noff] = 0x00;
+
+      off++;
+      noff++;
       break;
+    }
 
     switch (c & 0xc0) {
+      // Uncompressed label
       case 0x00: {
+        // Label size can not exceed maximum
         if (c > HSK_DNS_MAX_LABEL)
           return -1;
 
+        // Label size can not exceed length of data we are parsing
         if (off + c > data_len)
-          return -1; // EOF
-
-        if (noff + c + 1 > max)
           return -1;
 
-        int j;
-        for (j = off; j < off + c; j++) {
-          uint8_t b = data[j];
-
-          // Hack because we're
-          // using c-strings.
-          if (b == 0x00)
-            b = 0xff;
-
-          // This allows for double-dots
-          // too easily in our design.
-          if (b == 0x2e)
-            b = 0xfe;
-
-          if (name)
-            name[noff] = b;
-
-          noff += 1;
-        }
+        // Label can not extend full name past maximum including length byte
+        // and resolved compression pointers.
+        if (noff + c + 1 > HSK_DNS_MAX_NAME)
```

Comment:
I think here is similar. This only ensures there's enough space for the rest of the label, but not the `0x00`.

This will make last test case we already have in `record_read_vectors.h`:
```c
  {
    HSK_DNS_CNAME,
    "123456789012345678901234567890123456789012345678901234567890123.123456789012345678901234567890123456789012345678901234567890123.123456789012345678901234567890123456789012345678901234567890123.12345678901234567890123456789012345678901234567890123456789012.",
    "",
    "\x3f""123456789012345678901234567890123456789012345678901234567890123\x3f""123456789012345678901234567890123456789012345678901234567890123\x3f""123456789012345678901234567890123456789012345678901234567890123\x3e""12345678901234567890123456789012345678901234567890123456789012\x00",
    256,
    false,
    "\x3f""123456789012345678901234567890123456789012345678901234567890123\x3f""123456789012345678901234567890123456789012345678901234567890123\x3f""123456789012345678901234567890123456789012345678901234567890123\x3e""12345678901234567890123456789012345678901234567890123456789012\x00",
    256
  }
```

To here earlier.

Reply:
fixed in f5200ab1251950b897e8ef3f71f5a11737aeaa5a